File: bilfind_app/lib/models/post_model.dart
import 'package:equatable/equatable.dart';
import 'package:flutter/foundation.dart';

class PostModel extends Equatable {
  PostModel({
    required this.id,
    required this.userId,
    required this.title,
    required this.content,
    required this.images,
    this.price,
    required this.createdAt,
    required this.type,
    required this.ownerName,
    required this.ownerEmail,
    required this.ownerDepartment,
    required this.ownerUserId,
    this.favCount = 0,
    this.ownerPhoto,
    this.isMock = false,
    this.mockImages = const [],
  });
  late String id;
  late String userId;
  late String title;
  late int favCount;
  late String content;
  late List<String> images;
  List<Uint8List> mockImages = [];
  late double? price;
  late DateTime createdAt;
  late String type;
  late String ownerName;
  late String ownerEmail;
  late String ownerDepartment;
  late String? ownerPhoto;
  late String ownerUserId;
  late bool isMock;

  PostModel.fromJson(Map<String, dynamic> json) {
    id = json['id'];
    userId = json['userId'];
    title = json['title'];
    content = json['content'];
    images = json["images"] != null
        ? List.castFrom<dynamic, String>(json['images'])
        : [];
    price = json["price"];
    favCount = json["favCount"] ?? 0;
    createdAt = DateTime.parse(json['createdAt']).toLocal();
    type = json['type'];
    ownerName = json['ownerName'];
    ownerEmail = json['ownerEmail'];
    ownerDepartment = json['ownerDepartment'];
    ownerUserId = json['ownerUserId'];
    ownerPhoto = json['ownerPhoto'];
    isMock = false;
    mockImages = [];
  }

  @override
  List<Object?> get props => [id];
}

File: backend/src/models/post-model.ts
import { Expose, Transform } from "class-transformer";
import { IsEnum } from "class-validator";
import { ObjectId } from "mongodb";
import { User } from "./user-model";
import { Mapper } from "../utils/mapper";
import { Departments } from "../utils/enums";

export enum PostType {
  SALE = "SALE",
  BORROW = "BORROW",
  DONATION = "DONATION",
  FOUND = "FOUND",
  REQUEST = "REQUEST",
  LOST = "LOST",
  RENT = "RENT",
  FORUM = "FORUM"
}

export enum PostStatus {
  ACTIVE = "ACTIVE",
  DELETED = "DELETED",
  BANNED = "BANNED",
}

export enum PostCategory {
  CLOTHES = "CLOTHES",
  BOOKS = "BOOKS",
  ELECTRONIC = "ELECTRONIC",
}

export class PostModel {
  @Transform((value) => value.obj._id.toString())
  @Expose()
  _id?: ObjectId;

  @Expose()
  userId: string;

  @Expose()
  title: string;

  @Expose()
  content: string;

  @Expose()
  images?: string[];

  @Expose()
  price?: number;

  @Expose()
  createdAt: Date;

  @Expose()
  isDeleted: boolean;

  @Expose()
  @IsEnum(PostType)
  type: PostType;

  @Expose()
  @IsEnum(PostStatus)
  status: PostStatus;

  @Expose()
  department?: Departments;

  @Expose()
  category?: PostCategory;

  @Expose()
  favCount: number;
}

export class PostResponseDTO {
  @Expose()
  id: String;

  @Expose()
  userId: string;

  @Expose()
  title: string;

  @Expose()
  content: string;

  @Expose()
  images?: string[];

  @Expose()
  price?: number;

  @Expose()
  createdAt: Date;

  @Expose()
  @IsEnum(PostType)
  type: PostType;

  @Expose()
  ownerPhoto?: string;

  @Expose()
  ownerName: string;

  @Expose()
  ownerEmail: string;

  @Expose()
  ownerDepartment: Departments;

  @Expose()
  ownerUserId: String;

  @Expose()
  @IsEnum(PostStatus)
  status: PostStatus;

  @Expose()
  department?: Departments;

  @Expose()
  category?: PostCategory;

  @Expose()
  favCount: number;
}

export const mapToPostResponseDTO = (postModel: PostModel, user: User): PostResponseDTO => {
  return Mapper.map(PostResponseDTO, {
    ...postModel,
    id: postModel._id!.toString(),
    ownerPhoto: user.profilePhoto,
    ownerName: user.name + " " + user.familyName,
    ownerDepartment: user.departmant,
    ownerEmail: user.email,
    ownerUserId: user._id!.toString(),
  });
};
